File: DataStructures/Stacks/atcoder/MonotonicStack/A60/A60.ts
import * as fs from 'fs';

function main() {
    const input = fs.readFileSync('/dev/stdin', 'utf8').trim().split(/\s+/);
    const N = parseInt(input[0]);
    const A = input.slice(1).map(Number);

    const stack: number[] = []; // インデックス（0-based）を保持
    const result: number[] = new Array(N).fill(-1);

    for (let i = 0; i < N; i++) {
        // スタックの株価が今の株価以下なら捨てる（意味がない）
        while (stack.length > 0 && A[stack[stack.length - 1]] <= A[i]) {
            stack.pop();
        }

        // 残っていれば、直近の大きい株価の日が起算日
        if (stack.length > 0) {
            result[i] = stack[stack.length - 1] + 1; // 出力は1-basedに
        }

        // 今の日をスタックに積む
        stack.push(i);
    }

    console.log(result.join(' '));
}

main();

File: DataStructures/Stacks/atcoder/MonotonicStack/A60/A60.py
def main():
    import sys
    input = sys.stdin.read

    data = list(map(int, input().split()))
    N = data[0]
    A = data[1:]

    stack = []  # インデックスを保持
    result = [-1] * N

    for i in range(N):
        # 今の株価より小さいまたは同じ株価は使えない → pop
        while stack and A[stack[-1]] <= A[i]:
            stack.pop()

        # 残っていればそのインデックスが起算日（1-basedに直す）
        if stack:
            result[i] = stack[-1] + 1

        # 現在のインデックスをスタックに積む
        stack.append(i)

    print(' '.join(map(str, result)))

if __name__ == "__main__":
    main()
